File: flutter/library/lib/src/tsp_score_rule_grid.dart
import 'package:flutter/material.dart';
import 'star_planet_theme.dart';

class TspScoreRule {
  const TspScoreRule({
    this.id,
    this.name = '',
    this.icon,
    this.value = 0,
    this.count = 0,
    this.dailyLimit,
  });

  final Object? id;
  final String name;
  final String? icon;
  final num value;
  final num count;
  final num? dailyLimit;
}

/// Score rule cards with optional +1.
class TspScoreRuleGrid extends StatelessWidget {
  const TspScoreRuleGrid({
    super.key,
    this.rules = const [],
    this.columns = 'auto',
    this.variant = 'default',
    this.disabled = false,
    this.theme = StarPlanetTheme.sky,
    this.onIncrement,
  });

  final List<TspScoreRule> rules;
  final Object columns;
  final String variant;
  final bool disabled;
  final StarPlanetTheme theme;
  final ValueChanged<TspScoreRule>? onIncrement;

  int get _colCount {
    if (columns == 1 || columns == '1') return 1;
    if (columns == 2 || columns == '2') return 2;
    return 2;
  }

  @override
  Widget build(BuildContext context) {
    final readOnly = variant == 'readOnly' || disabled;
    final colCount = _colCount;

    return LayoutBuilder(
      builder: (context, constraints) {
        final gap = 12.0;
        final useSingle = columns == 1 || columns == '1' || (columns == 'auto' && constraints.maxWidth < 481);
        final cols = useSingle ? 1 : colCount;
        final cardWidth = cols == 1 ? constraints.maxWidth : (constraints.maxWidth - gap) / 2;

        return Wrap(
          spacing: gap,
          runSpacing: gap,
          children: rules.map((rule) {
            final value = rule.value;
            final count = rule.count;
            final limit = rule.dailyLimit;
            final atLimit = limit != null && count >= limit;
            final valueText = value > 0 ? '+$value' : '$value';
            final meta = limit == null ? '已记 $count' : '已记 $count/$limit';

            return SizedBox(
              width: cardWidth,
              child: Container(
                padding: const EdgeInsets.all(12),
                decoration: BoxDecoration(
                  color: theme.surfaceRaised,
                  borderRadius: BorderRadius.circular(18),
                  border: Border.all(color: theme.borderDefault),
                ),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Row(
                      children: [
                        if (rule.icon != null) ...[
                          Text(rule.icon!),
                          const SizedBox(width: 6),
                        ],
                        Expanded(
                          child: Text(
                            rule.name,
                            style: TextStyle(color: theme.textPrimary, fontWeight: FontWeight.w800, fontSize: 13),
                          ),
                        ),
                      ],
                    ),
                    const SizedBox(height: 8),
                    Text(
                      valueText,
                      style: TextStyle(
                        color: value >= 0 ? theme.success : theme.danger,
                        fontSize: 22,
                        fontWeight: FontWeight.w900,
                      ),
                    ),
                    const SizedBox(height: 4),
                    Text(meta, style: TextStyle(color: theme.textTertiary, fontSize: 12)),
                    if (!readOnly) ...[
                      const SizedBox(height: 8),
                      Opacity(
                        opacity: (atLimit || disabled) ? .4 : 1,
                        child: GestureDetector(
                          onTap: (atLimit || disabled) ? null : () => onIncrement?.call(rule),
                          child: Container(
                            width: double.infinity,
                            constraints: const BoxConstraints(minHeight: 44),
                            alignment: Alignment.center,
                            decoration: BoxDecoration(
                              color: theme.selectedFill,
                              borderRadius: BorderRadius.circular(999),
                              border: Border.all(color: theme.borderDefault),
                            ),
                            child: Text(
                              '+1',
                              style: TextStyle(color: theme.brandDark, fontWeight: FontWeight.w900),
                            ),
                          ),
                        ),
                      ),
                    ],
                  ],
                ),
              ),
            );
          }).toList(),
        );
      },
    );
  }
}

File: react-native/library/src/starPlanet/components/TspScoreRuleGrid.js
import React from 'react';
import { Pressable, StyleSheet, Text, View } from 'react-native';
import { withTheme } from '../utils/shared';

/**
 * TspScoreRuleGrid – Score rule cards with optional +1.
 */
export function TspScoreRuleGrid({
  rules = [],
  columns = 'auto',
  variant = 'default',
  disabled = false,
  theme,
  onIncrement,
}) {
  const t = withTheme(theme);
  const readOnly = variant === 'readOnly' || disabled;
  const colCount = columns === 1 || columns === '1' ? 1 : 2;

  return (
    <View style={styles.grid}>
      {rules.map((rule) => {
        const value = Number(rule.value) || 0;
        const count = Number(rule.count) || 0;
        const limit = rule.dailyLimit == null ? null : Number(rule.dailyLimit);
        const atLimit = limit != null && count >= limit;
        return (
          <View
            key={String(rule.id ?? rule.name)}
            style={[
              styles.card,
              {
                width: colCount === 1 ? '100%' : '48%',
                backgroundColor: t.surfaceRaised,
                borderColor: t.borderDefault,
              },
            ]}
          >
            <View style={styles.head}>
              {rule.icon ? <Text>{rule.icon}</Text> : null}
              <Text style={[styles.name, { color: t.textPrimary }]} numberOfLines={1}>
                {rule.name || ''}
              </Text>
            </View>
            <Text
              style={[
                styles.value,
                { color: value >= 0 ? t.success : t.danger },
              ]}
            >
              {value > 0 ? `+${value}` : String(value)}
            </Text>
            <Text style={[styles.meta, { color: t.textTertiary }]}>
              {limit == null ? `已记 ${count}` : `已记 ${count}/${limit}`}
            </Text>
            {readOnly ? null : (
              <Pressable
                disabled={atLimit || disabled}
                onPress={() => onIncrement?.(rule)}
                style={[
                  styles.add,
                  {
                    backgroundColor: t.brandSubtle ?? t.selectedFill,
                    borderColor: t.borderDefault,
                  },
                  (atLimit || disabled) && styles.disabled,
                ]}
              >
                <Text style={[styles.addText, { color: t.brandPrimary }]}>+1</Text>
              </Pressable>
            )}
          </View>
        );
      })}
    </View>
  );
}

const styles = StyleSheet.create({
  grid: {
    flexDirection: 'row',
    flexWrap: 'wrap',
    gap: 12,
    justifyContent: 'space-between',
  },
  card: {
    borderWidth: 1,
    borderRadius: 18,
    padding: 14,
    gap: 6,
  },
  head: { flexDirection: 'row', alignItems: 'center', gap: 6 },
  name: { flex: 1, fontWeight: '800', fontSize: 14 },
  value: { fontSize: 22, fontWeight: '900' },
  meta: { fontSize: 12 },
  add: {
    marginTop: 4,
    alignSelf: 'flex-start',
    borderWidth: 1,
    borderRadius: 999,
    paddingHorizontal: 12,
    paddingVertical: 6,
  },
  addText: { fontWeight: '900', fontSize: 13 },
  disabled: { opacity: 0.4 },
});
